### PR TITLE
Adds collective_noun Fields Back to SiteConfig Model

### PR DESCRIPTION
**Pull request by juliannatetreault**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR adds the `collective_noun` and `collective_noun_disabled` columns back to the `SiteConfig` model after running into a `data_update` script failure in another PR. This is necessary so that when running the `data_update` script responsible for updating the `community_name`, we have access to the `collective_noun` and `collective_noun_disabled` methods. Additionally, this PR adds a `data_update` script that conditionally updates `SiteConfig.community_name` _if_ `SiteConfig.collective_noun_disabled` returns `false`. If, however, the forem has `collective_noun_disabled` set to `true`, then the `community_name` will remain as-is and untouched.

## Related Tickets & Documents
Relates to PRs #11846 , PR #11973 , and PR #12054 

## QA Instructions, Screenshots, Recordings
**To QA these changes, first ensure that all checks pass and the Travis build is green, then**:
- Run the `data_update` script manually and ensure that your Community's name updates accordingly

### UI accessibility concerns?
N/A

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
The `data_update` script included in this PR needs to successfully run!

## [optional] What gif best describes this PR or how it makes you feel?

The Office: Days of Nonsense

### CODE REVIEW

File: spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+require Rails.root.join(
+  "lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb",
+)
+
+describe DataUpdateScripts::AppendCollectiveNounToCommunityName do
+  context "when collective_noun_disabled is false" do
+    it "appends the collective_noun to the community_name" do
+      described_class.new.run
+      expect(SiteConfig.collective_noun_disabled).to eq(false)
+      expect(SiteConfig.community_name).to eq("DEV(local) Community")
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:11]: This continues to return `"DEV(local)"` rather than `"DEV(local) Community"` and after dropping into `pry` and trying a bunch of different things, I still cannot get this to work properly. It almost feels like I need a `reload` here, but I'm unable to use that on a `string` -- thoughts? Am I missing something? 😅 

cc @vaidehijoshi and @maestromac

File: lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module DataUpdateScripts
+  class AppendCollectiveNounToCommunityName
+    def run
+      return "#{SiteConfig.community_name} #{SiteConfig.collective_noun}" unless SiteConfig.collective_noun_disabled?
```
Comment (vaidehijoshi) [inline lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:4]: @juliannatetreault This doesn't actually _update_ the SiteConfig at all, it just returns a string. I think that's probably why the test isn't doing what you expect/want it to -- you need actually update the `SiteConfig.community_name` similar to what you [previously did](https://github.com/forem/forem/pull/11846/files#diff-620cde66dafdd0fed658093095e053565f368bee979f03ba83bc2eb9a2fe9dabR7) in your initial PR

File: spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  "lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb",
+)
+
+describe DataUpdateScripts::AppendCollectiveNounToCommunityName do
+  context "when collective_noun_disabled is false" do
+    it "appends the collective_noun to the community_name" do
+      SiteConfig.collective_noun = "Club"
+      described_class.new.run
+      expect(SiteConfig.collective_noun_disabled).to eq(false)
+      expect(SiteConfig.community_name).to eq("DEV(local) Club")
```
Comment (juliannatetreault) [inline spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:12]: I could also change this to mirror the code snippet below if folks think it's more readable and/or prefer it written this way:
```
      expect(SiteConfig.collective_noun_disabled).to eq(false)
      expect do
        described_class.new.run
      end.to change(SiteConfig, :community_name).from("DEV(local)").to("DEV(local) Club")
```
(The same goes for the test below!)

File: lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb (hunk 2 above)
Comment (juliannatetreault) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb]: I was totally overthinking this yesterday and looking at it with fresh eyes this morning made me 🤦‍♀️  --  thank you for the feedback!

File: spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb (hunk 3 above)
Comment (vaidehijoshi) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:12]: I like this a little better, and I think I've mostly seen this as a convention in our codebase!

File: spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require "rails_helper"
+require Rails.root.join(
+  "lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb",
+)
+
+describe DataUpdateScripts::AppendCollectiveNounToCommunityName do
+  context "when collective_noun_disabled is false" do
+    it "appends the collective_noun to the community_name" do
+      SiteConfig.collective_noun = "Club"
```
Comment (vaidehijoshi) [inline spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:9]: Rather than explicitly assigning a value to a field on `SiteConfig` here, I think we should opt to use Rspec's [method stubs](https://relishapp.com/rspec/rspec-mocks/v/2-99/docs/method-stubs) in this test and the one below, too.

By using a stub here, we can be sure that we're not accidentally ever changing the value of `SiteConfig.collective_noun`, which could leak into other tests or cause strange, confusing bugs in tests down the road.

In this test, that would look something like:
```ruby
allow(SiteConfig).to receive(:collective_noun).and_return("Club")
```

Comment (juliannatetreault) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:9]: Ahh, that's great! I could also just lean into the default `collective_noun`, which is "Community"--I set `collective_noun` to "Club" here just to make the test more explicit for folks who look at this code in the future. Thoughts @vaidehijoshi ?

Comment (vaidehijoshi) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb:9]: That would work!

But, there's one thing to keep in mind here, which can a bit obtuse until you think about how this test might be interpreted in the future. There is one _benefit_ of being explicit in the test, like the way you currently have it. The benefit is that if the (default) value you are testing against changes down the road (aka someone goes and changes [this line](https://github.com/forem/forem/pull/12055/files#diff-68da686a0f7348bf9a2c59e669043f21019ab8dfb6363cea6daa534c5c8296d7R58)), then this test would start failing. Sometimes, you _want_ your test to fail in that way, and relying on the default makes sense.

But this test isn't really testing the default value of `community_club`; it's testing the behavior of the _data update script_. So, if you were _not_ explicit about the value of this field in this test and if this fails down the road, it could be confusing (at first) to the person making the change and they'd have to figure out that the test was failing because it was _implicitly_ depending upon + testing the default value of the SiteConfig's field. By being explicit here, you are actually avoiding that scenario!

TL;DR, I would vote to be explicit, just as you have it set up already 😉

Comment (juliannatetreault) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb]: Thank you so much for the thorough explanation! 😄 

I just pushed up changes that incorporated method stubs in the spec before seeing this comment @vaidehijoshi . 😂 I think the stubs sound like a safer way forward, even if they're not as explicit. That being said, would it be helpful if I added a comment to the spec now that I've changed it to give future contributors/us a bit more context?

Comment (vaidehijoshi) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb]: > I think the stubs sound like a safer way forward, even if they're not as explicit.

@juliannatetreault I think maybe there is a little confusion here: the stubs don't actually have anything to do with the explicitness that I was talking about in my comment. What makes this test explicit is when you declare that `collective_noun` is set to return a value of the string `"Club"`:
```ruby
# explicitly setting the collective_noun by using a method stub
allow(SiteConfig).to receive(:collective_noun).and_return("Club")
```

To be clear, what you had in your initial commit was _also explicit_, even though it didn't use a stub:
```ruby
# explicitly setting the collective_noun by setting the value directly, which is less safe
SiteConfig.collective_noun = "Club"
```
Both of these are explicit because you are explicitly setting the value of `collective_noun` (either with a method stub, or directly) in the body of the test. When someone reads either of those two variations above,  

The _implicitness_ I was talking about would have only applied if you had relied on the default value, like you were describing in your earlier comment:
> Alternatively, I could also just lean into the default value for collective_noun, which is "Community"

So, if you had written your test like this:
```ruby
described_class.new.run
expect(SiteConfig.collective_noun_disabled).to eq(false)
expect(SiteConfig.community_name).to eq("DEV(local) Community")
```

then _that_ would have been implicit, because you would have been implicitly relying on the fact that the default value of `collective_noun` is `"Community"`. 

Hopefully that explanation makes more sense? :) Let me know if any of it is still confusing. Given all of that, I don't think you need to add a comment here because the test is being explicit while using a stub 👍🏽

Comment (juliannatetreault) [reply spec/lib/data_update_scripts/append_collective_noun_to_community_name_spec.rb]: OH! I misread your comment, @vaidehijoshi -- apologies. 😊 I'll leave everything as-is then.

File: app/models/site_config.rb [hunk 5]
```diff
@@ -55,6 +55,8 @@ class SiteConfig < RailsSettings::Base
   # Community Content
   field :community_name, type: :string, default: ApplicationConfig["COMMUNITY_NAME"] || "New Forem"
   field :community_emoji, type: :string, default: "🌱"
+  field :collective_noun, type: :string, default: "Community"
+  field :collective_noun_disabled, type: :boolean, default: false
```
Comment (vaidehijoshi) [inline app/models/site_config.rb:59]: We might want to add a comment next to these 2 fields explaining that we don't use these across the app elsewhere right now and that they are only here temporarily/will be removed when this issue is fully resolved

**Review by vaidehijoshi — APPROVED** [review]
Works great locally, and thank you for adding a test! 🚀

File: app/models/site_config.rb (hunk 5 above)
Comment (juliannatetreault) [reply app/models/site_config.rb:61]: Great call! Done ✔️

File: lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module DataUpdateScripts
+  class AppendCollectiveNounToCommunityName
+    def run
+      return if SiteConfig.collective_noun_disabled
+
+      SiteConfig.community_name = "#{SiteConfig.community_name} #{SiteConfig.collective_noun}"
```
Comment (maestromac) [inline lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: Is there a possibility for collective_noun to be nil?

**Review by maestromac — APPROVED** [review]
LGTM!

File: lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb (hunk 6 above)
Comment (juliannatetreault) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: Great question! I just tested this out in the console and it seems like there is a possibility that `collective_noun` could be set to `nil`, but I don't know if this is actually the case for any communities. That being said, I could adjust line 4 to read:
```
return if SiteConfig.collective_noun_disabled || SiteConfig.collective_noun.nil?
```
Or instead, maybe I could add a `nil` guard to line 6 like so:
```
#{SiteConfig&.collective_noun}"
```

Thoughts @maestromac ?

Comment (maestromac) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: That works! you should use `!SiteConfig.collective_noun.present?` instead though since that will account for empty string

Comment (juliannatetreault) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: @maestromac What if I changed it to the code snippet below to account for empty strings, `nil` values, and empty arrays/hashes?
```
SiteConfig.collective_noun.blank?
```

Comment (maestromac) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: great idea!

Comment (juliannatetreault) [reply lib/data_update_scripts/20201228194641_append_collective_noun_to_community_name.rb:6]: Done! I'll merge/deploy pending a green and passing build. 😄
